### PR TITLE
Tweak: Swiper loads css only when needed [ED-15015]

### PR DESCRIPTION
**Pull request by mugurelLupu**
## PR Checklist

- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

### CODE REVIEW

File: includes/frontend.php [hunk 1]
```diff
@@ -638,7 +638,7 @@ public function enqueue_styles() {
 
 			wp_enqueue_style( 'elementor-frontend' );
 
-			wp_enqueue_style( 'swiper' );
+			//wp_enqueue_style( 'swiper' );  We can remove this in version 3.26
```
Comment (hein-obox) [inline includes/frontend.php:641]: Please add this comment to the ED-15471 task.

```suggestion
			// TODO: Remove in version 3.26.0 [ED-15471]
			//wp_enqueue_style( 'swiper' );
```

File: core/page-assets/loader.php [hunk 2]
```diff
@@ -35,6 +35,11 @@ private function init_assets() {
 					'version' => ELEMENTOR_VERSION,
 					'dependencies' => [],
 				],
+				'swiper' => [
+					'src' => $this->get_css_assets_url( 'swiper',  Plugin::$instance->experiments->is_feature_active( 'e_swiper_latest' ) ? 'assets/lib/swiper/v8/css/' : 'assets/lib/swiper/css/' ),
```
Comment (hein-obox) [inline core/page-assets/loader.php:39]: Please extract to a function, like it was done inside frontend.php, for a more readable code.

File: includes/elements/section.php [hunk 3]
```diff
@@ -552,6 +552,22 @@ protected function register_controls() {
 				'fields_options' => [
 					'background' => [
 						'frontend_available' => true,
+						'assets' => [
+							'styles' => [
+								[
+									'name' => 'swipper',
```
Comment (hein-obox) [inline includes/elements/section.php:558]: I am not sure if this approach is working?

```suggestion
									'name' => 'swiper',
```

File: includes/frontend.php (hunk 1 above)
Comment (hein-obox) [reply includes/frontend.php:641]: Can you implement this suggestion?

File: includes/elements/container.php [hunk 4]
```diff
@@ -734,6 +758,22 @@ protected function register_background_controls() {
 				'selectors' => [
 					'{{WRAPPER}}' => '--background-transition: {{SIZE}}s;',
 				],
+//					'assets' => [
```
Comment (hein-obox) [inline includes/elements/container.php:761]: Please clean up.

File: includes/elements/container.php [hunk 5]
```diff
@@ -1172,22 +1212,22 @@ protected function register_shape_dividers_controls() {
...
-											'name' => $base_control_key,
-											'operator' => '!==',
-											'value' => '',
-										],
-									],
-								],
-							],
-						],
-					],
+//					'assets' => [
```
Comment (hein-obox) [inline includes/elements/container.php:1215]: Please restore.

File: includes/elements/section.php [hunk 6]
```diff
@@ -552,6 +552,22 @@ protected function register_controls() {
 				'fields_options' => [
 					'background' => [
 						'frontend_available' => true,
+						'assets' => [
```
Comment (hein-obox) [inline includes/elements/section.php:555]: I don't think this is working. 

Has it been tested?

Why don't we have the same implementation as the container.php?

Comment (hein-obox) [reply includes/elements/section.php:555]: You can test the Sections, by deactivating the Flex Container experiment.

File: includes/elements/container.php [hunk 7]
```diff
@@ -734,6 +758,22 @@ protected function register_background_controls() {
 				'selectors' => [
 					'{{WRAPPER}}' => '--background-transition: {{SIZE}}s;',
 				],
+				'assets' => [
```
Comment (hein-obox) [inline includes/elements/container.php:761]: Do we need this code? 

I would assume that we only to conditionally load the swiper styling once.

@mugurelLupu

File: includes/elements/container.php (hunk 5 above)
Comment (hein-obox) [reply includes/elements/container.php:1215]: @mugurelLupu Not updated yet?

File: core/page-assets/loader.php [hunk 8]
```diff
@@ -35,11 +35,23 @@ private function init_assets() {
...
+				],
 			],
 			'scripts' => [],
 		];
 	}
 
+	private function getSwiperPath() {
+		return Plugin::$instance->experiments->is_feature_active( 'e_swiper_latest' ) ? 'assets/lib/swiper/v8/css/' : 'assets/lib/swiper/css/';
+	}
+	private function getSwiperVersion() {
```
Comment (hein-obox) [inline core/page-assets/loader.php:51]:
```suggestion

	private function getSwiperVersion() {
```

File: includes/frontend.php [hunk 9]
```diff
@@ -648,7 +648,8 @@ public function enqueue_styles() {
 
 			wp_enqueue_style( 'elementor-frontend' );
 
-			wp_enqueue_style( 'swiper' );
+			// TODO: Remove in version 3.26.0 [ED-15471]
+			//wp_enqueue_style( 'swiper' );
```
Comment (hein-obox) [inline includes/frontend.php:652]: I think we can uncomment this now.

File: tests/playwright/pages/editor-page.ts [hunk 10]
```diff
@@ -2,7 +2,7 @@ import { addElement, getElementSelector } from '../assets/elements-utils';
 import { expect, type Page, type Frame, type TestInfo } from '@playwright/test';
 import BasePage from './base-page';
 import EditorSelectors from '../selectors/editor-selectors';
-import _path from 'path';
+import _path, {resolve} from 'path';
```
Comment (hein-obox) [inline tests/playwright/pages/editor-page.ts:5]:
```suggestion
import _path, { resolve } from 'path';
```

File: tests/playwright/sanity/modules/container/container-1.test.ts [hunk 11]
```diff
@@ -26,6 +26,31 @@ test.describe( 'Container tests @container', () => {
 		await page.close();
 	} );
 
+	test('Background slideshow', async ({ page, apiRequests }, testInfo) => {
```
Comment (hein-obox) [inline tests/playwright/sanity/modules/container/container-1.test.ts:29]:
```suggestion
	test( 'Background slideshow', async ( { page, apiRequests }, testInfo ) => {
```

File: tests/playwright/sanity/modules/container/container-1.test.ts [hunk 12]
```diff
@@ -26,6 +26,31 @@ test.describe( 'Container tests @container', () => {
...
+		const wpAdmin = new WpAdminPage(page, testInfo, apiRequests);
```
Comment (hein-obox) [inline tests/playwright/sanity/modules/container/container-1.test.ts:30]:
```suggestion
		const wpAdmin = new WpAdminPage( page, testInfo, apiRequests );
```

File: tests/playwright/sanity/modules/container/container-1.test.ts [hunk 13]
```diff
@@ -26,6 +26,31 @@ test.describe( 'Container tests @container', () => {
...
+		const editor = await wpAdmin.openNewPage();
+		const containerId = await editor.addElement( { elType: 'container' }, 'document' );
+		await editor.addWidget( 'heading', containerId );
+		await editor.selectElement( containerId );
+
+		await editor.closeNavigatorIfOpen();
+		await editor.openPanelTab('style');
```
Comment (hein-obox) [inline tests/playwright/sanity/modules/container/container-1.test.ts:37]:
```suggestion
		await editor.openPanelTab( 'style' );
```

File: tests/playwright/sanity/modules/container/container-1.test.ts [hunk 14]
```diff
@@ -26,6 +26,31 @@ test.describe( 'Container tests @container', () => {
...
+		await editor.openSection('section_background');
```
Comment (hein-obox) [inline tests/playwright/sanity/modules/container/container-1.test.ts:38]:
```suggestion
		await editor.openSection( 'section_background' );
```

File: includes/elements/section.php [hunk 15]
```diff
@@ -557,6 +557,29 @@ protected function register_controls() {
 			]
 		);
 
+		$this->add_control(
+		'handle_slideshow_asset_loading',
```
Comment (hein-obox) [inline includes/elements/section.php:561]:
```suggestion
			'handle_slideshow_asset_loading',
```

**Review by hein-obox — DISMISSED** [review]

File: includes/widgets/image-carousel.php [hunk 16]
```diff
@@ -88,9 +88,13 @@ protected function is_dynamic_content(): bool {
...
-		return $this->load_widgets_styles_in_head()
-			? [ 'widget-image-carousel' ]
-			: [];
+		$assets = [ 'e-swiper', 'swiper' ];
+
+		if ( $this->load_widgets_styles_in_head() ) {
+			$assets[] = 'widget-image-carousel';
+		}
+
+		return $assets;
```
Comment (rami-elementor) [inline includes/widgets/image-carousel.php:97]: It's a minor change, but this is the convention:

```suggestion
		$style_depends = [ 'e-swiper', 'swiper' ];

		if ( $this->load_widgets_styles_in_head() ) {
			$style_depends[] = 'widget-image-carousel';
		}

		return $style_depends;
```

File: tests/playwright/sanity/core/admin/mytest.test.ts
Comment (rami-elementor) [inline tests/playwright/sanity/core/admin/mytest.test.ts:1]: An empty file?

Comment (hein-obox) [reply tests/playwright/sanity/core/admin/mytest.test.ts:1]: @rami-elementor Fixed

File: core/page-assets/loader.php [hunk 17]
```diff
@@ -35,11 +35,29 @@ private function init_assets() {
 					'version' => ELEMENTOR_VERSION,
 					'dependencies' => [],
 				],
+				'e-swiper' => [
+					'src' => $this->get_css_assets_url( 'e-swiper', 'assets/css/conditionals/' ),
+					'version' => ELEMENTOR_VERSION,
+					'dependencies' => [],
```
Comment (rami-elementor) [inline core/page-assets/loader.php:41]: Not sure, but shouldn't we add `swiper` as a dependency for `e-swiper`?

File: includes/frontend.php [hunk 18]
```diff
@@ -540,6 +540,13 @@ public function register_styles() {
 			ELEMENTOR_VERSION
 		);
 
+		wp_register_style(
+			'e-swiper',
+			$this->get_css_assets_url( 'e-swiper', 'assets/css/conditionals/' ),
+			[],
```
Comment (rami-elementor) [inline includes/frontend.php:546]: Same here? should we add `swiper` dependency for `e-swiper`?

File: includes/frontend.php [hunk 19]
```diff
@@ -541,10 +541,10 @@ public function register_styles() {
...
 
 		wp_register_style(
-			'swiper',
-			$this->get_css_assets_url( 'swiper', $this->e_swiper_asset_path . 'css/' ),
-			[],
-			$this->e_swiper_version
+			'e-swiper',
+			$this->get_css_assets_url( 'e-swiper', 'assets/css/conditionals/' ),
+			[ 'swiper' ],
+			ELEMENTOR_VERSION
```
Comment (rami-elementor) [inline includes/frontend.php:547]: Question: why did you deleted the original `swiper` ?

File: core/page-assets/loader.php (hunk 17 above)
Comment (hein-obox) [reply core/page-assets/loader.php]: I love it.

File: includes/frontend.php (hunk 18 above)
Comment (hein-obox) [reply includes/frontend.php]: Implemented

File: includes/frontend.php (hunk 19 above)
Comment (hein-obox) [reply includes/frontend.php:547]: Restored.

**Review by rami-elementor — APPROVED** [review]

**Review by louiswol94 — APPROVED** [review]

**Pull request merged**